Frameworks/CoreGraphics/CGFont.mm

```diff
@@ -16,615 +16,134 @@
-
-        curWidth = ((float)(penX)) / 64.0f;
-        if (curWidth > width && curChar != ' ') {
-            if (lastPossibleBreakPos != -1) {
-                //  We must now perform a soft break
-                state->lineLen = &state->chars[lastPossibleBreakPos] - state->lineStart;
-                state->curIndex = lastPossibleBreakPos + 1;
-
-                if (curChar == ' ') {
-                    //  Do a hard line break
-                    state->hitLinebreak = true;
-                }
-                if (sizeOut) {
-                    sizeOut->width = ((float)lastPossibleBreakWidth) / 64.0f;
-                }
-            } else {
-                //  Draw as many characters as we can
-                if (state->lineStart != &state->chars[state->curIndex]) {
-                    state->lineLen = &state->chars[state->curIndex] - state->lineStart;
-                    penX -= advance.x;
-
-                    if (sizeOut) {
-                        sizeOut->width = ((float)penX) / 64.0f;
-                    }
-                    break;
-                } else {
-                    //  Draw at least one char
-                    state->lineLen = 1;
-                    state->curIndex++;
-
-                    if (sizeOut) {
-                        sizeOut->width = ((float)penX) / 64.0f;
-                    }
-                    break;
-                }
-            }
-            break;
-        }
-
-        state->curIndex++;
-    }
-
-    //  Did not wrap
-    if (state->lineLen == -1) {
-        state->lineLen = &state->chars[state->count] - state->lineStart;
-        if (sizeOut) {
-            sizeOut->width = ((float)penX) / 64.0f;
-        }
-    }
-
-    return ret;
+static void __CGFontInit(CFTypeRef cf) {
+    CGFontRef font = (CGFontRef)cf;
+    // ComPtr needs to be manually new'd, to avoid calling Create functions to an incomplete out-pointer
+    struct __CGFont* mutableFont = const_cast<struct __CGFont*>(font);
+    new (std::addressof(mutableFont->_dwriteFontFace)) ComPtr<IDWriteFontFace>();
 }
 
-bool CGFontWrap(id font, float size, CGFontWrapState* state, float width, CGSize* sizeOut) {
-    if (!font) {
-        if (sizeOut) {
-            sizeOut->width = 0;
-            sizeOut->height = 0;
-        }
-        return false;
-    }
-
-    _CGFontLock();
-
-    //  Get the font
-    FT_Face face = (FT_Face)(DWORD)[font _sizingFontHandle];
-    CGFontSetFTFontSize(font, face, size);
-
-    if (sizeOut && face) {
-        sizeOut->width = 0;
-        sizeOut->height = ((float)(face->size->metrics.ascender - face->size->metrics.descender)) * spacing / 64.0f;
-    }
-
-    bool ret = CGFontWrapFunc(state, width, sizeOut, freeTypeGetCharWidth, face);
-
-    _CGFontUnlock();
-
-    return ret;
+static void __CGFontDeallocate(CFTypeRef cf) {
+    CGFontRef font = (CGFontRef)cf;
+    // ComPtr needs to be manually destructed, since CFTypes do a memset(0) on dealloc
+    font->_dwriteFontFace.~ComPtr();
 }
```

Comment:
This is necessary for integration with CF while having C++ members.

---
In reply to: [84554973](https://github.com/Microsoft/WinObjC/pull/1145#discussion_r84554973) [](ancestors = 84554973,84415796,83340447)